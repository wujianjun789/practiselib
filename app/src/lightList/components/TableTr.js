/**
 * Created by Azrael on 2017/9/15.
 */
import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
/**
 * 这里是注释
 * 这里是注释
 * 这里是注释
 * 这里是注释
 */
export default class TableTr extends PureComponent {
    static propTypes = {
        activeId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        columns: PropTypes.arrayOf(PropTypes.shape({
            accessor: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired,
            title: PropTypes.string.isRequired
        }) ),
        willMountFuncs: PropTypes.oneOfType([
            PropTypes.arrayOf(PropTypes.func),
            PropTypes.func
        ]),
        rowClick: PropTypes.func
    }

    constructor(props) {
        super(props);
        this.state = {
            data: {...props.data},
        };

        this.onClick = this.onClick.bind(this);
        this.updateData = this.updateData.bind(this);
    }

    componentWillMount() {
        this.mounted = true;
        const {willMountFuncs} = this.props;
        if (Array.isArray(willMountFuncs)) {
            willMountFuncs.forEach(func => {func(this.updateData)});
        } else if (typeof willMountFuncs == 'function') {
            willMountFuncs(this.updateData);
        }
    }

    componentWillUnmount() {
        this.mounted = false;
    }

    updateData(data) {
        this.mounted && this.setState({data: {...this.state.data, ...data}});
    }

    onClick() {
        this.props.rowClick && this.props.rowClick(this.state.data);
    }

    render() {
        const {data, data:{id}} = this.state;
        const {activeId, columns} = this.props;

        return (
            <tr className={(activeId || (typeof activeId == 'number' && activeId == 0) ) && activeId==id ? 'active':''} onClick={this.onClick}>
            {
                columns.map((item, index) => {
                    if(typeof item.accessor === 'function') {
                        return <td key={index}>{item.accessor(data)}</td>
                    } else if(typeof item.accessor === 'string') {
                        return <td key={index}>{data[item.accessor]}</td>
                    } else {
                        return <td key={index}></td>
                    }
                })
            }
            </tr>
        )
    }
}
