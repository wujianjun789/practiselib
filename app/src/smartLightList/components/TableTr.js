/**
 * Created by Azrael on 2017/9/15.
 */
import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
/**
 * 这里是注释
 * 这里是注释
 * 这里是注释
 * 这里是注释
 */
export default class TableTr extends PureComponent {
    static propTypes = {
        activeId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        columns: PropTypes.arrayOf(PropTypes.shape({
            accessor: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired,
            title: PropTypes.string.isRequired
        }) ),
        formatFunc: PropTypes.func,
        willMountFuncs: PropTypes.oneOfType([
            PropTypes.arrayOf(PropTypes.func),
            PropTypes.func
        ]),
        rowClick: PropTypes.func
    }

    constructor(props) {
        super(props);
        this.state = {
            data: {...props.data},
            activeId: props.activeId
        };

        this.onClick = this.onClick.bind(this);
        this.updateData = this.updateData.bind(this);
    }

    componentWillMount() {
        this.mounted = true;
        const {willMountFuncs} = this.props;
        if (Array.isArray(willMountFuncs)) {
            willMountFuncs.forEach(func => {func(this.updateData)});
        } else if (typeof willMountFuncs == 'function') {
            willMountFuncs(this.updateData);
        }
    }

    componentWillUnmount() {
        this.mounted = false;
    }

    updateData(data) {
        this.mounted && this.setState({data: {...this.state.data, ...data}});
    }

    onClick() {
        this.props.rowClick && this.props.rowClick(this.state.data);
    }

    render() {
        const {data, data:{id}} = this.state;
        const {activeId, columns, formatFunc} = this.props;

        let _data = {...data};
        if(typeof formatFunc == 'function') {
            formatFunc(_data);
         }

        return (
            <tr className={activeId && activeId==id ? 'active':''} onClick={this.onClick}>
            {
                columns.map((item, index) => {
                    if(typeof item.accessor === 'function') {
                        return <td key={index}>{item.accessor(_data)}</td>
                    } else if(typeof item.accessor === 'string') {
                        return <td key={index}>{_data[item.accessor]}</td>
                    } else {
                        return <td key={index}></td>
                    }
                })
            }
            </tr>
        )
    }
}