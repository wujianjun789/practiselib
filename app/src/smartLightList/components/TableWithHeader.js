/**
 * Created by Azrael on 2017/9/15.
 */
import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
/**
 * table container with thead
 */
export default class TableWithHeader extends PureComponent {
    static propTypes = {
        className: PropTypes.string,
        columns: PropTypes.arrayOf(PropTypes.shape({
            field: PropTypes.string.isRequired,
            title: PropTypes.string.isRequired
        }) ),
        children: PropTypes.arrayOf(PropTypes.node)
    }

    render() {
        const {className, columns, children} = this.props;
        return (
            <div className={`table-responsive ${className}`}>
            <table className="table table-hover">
                <thead>
                <tr>
                    {
                        columns.map((item, index)=> {
                            return <th key={index}>{item.title}</th>
                        })
                    }
                </tr>
                </thead>
                <tbody>
                    {children}
                </tbody>
            </table>
        </div>
        )
    }
}