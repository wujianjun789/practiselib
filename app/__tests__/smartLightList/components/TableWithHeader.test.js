import React from 'react';
import {shallow} from 'enzyme';
import renderer from 'react-test-renderer';
import TableWithHeader from '../../../src/smartLightList/components/TableWithHeader';
import TableTr from '../../../src/smartLightList/components/TableTr';

describe('<TableWithHeader />', () => {
	const columns = [
		{accessor: 'name', title: '设备名称'},
		{accessor(data){return data.value*10}, title: 'value*10'}
	];
	const dataList = [
		{ id: 0, name: 'test', value: 10},
		{ id: 1, name: 'test', value: 10}
	];

	it('default render', () => {
		const cmp = shallow(<TableWithHeader columns={columns}/>);
		const th = cmp.find('thead').find('th');
		expect(th.length).toBe(2);
		columns.forEach((item, index) => {
			expect(th.at(index).text()).toBe(item.title);
		});
	});

	it('render with className="test"', () => {
		const cmp = shallow(<TableWithHeader columns={columns} className="test"/>);
		expect(cmp.find('.table-responsive').hasClass('test')).toBeTruthy();
	});

	it('render with children', () => {
		const cmp = shallow(<TableWithHeader columns={columns}>
			{
				dataList.map(item => {
					return <TableTr key={item.id} data={item} columns={columns}/>
				})
			}
			</TableWithHeader>);

		const tr = cmp.find('TableTr');
		expect(tr.length).toBe(dataList.length);
		dataList.forEach((item, index) => {
			expect(tr.at(index).key()).toEqual(''+item.id);
			expect(tr.at(index).prop('data')).toEqual(item);
			expect(tr.at(index).prop('columns')).toEqual(columns);
		});
	});

	it('default render snapshot', () => {
		const cmp = renderer.create(<TableWithHeader columns={columns}/>);
		const tree = cmp.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('render with className="test" snapshot', () => {
		const cmp = renderer.create(<TableWithHeader columns={columns} className="test"/>);
		const tree = cmp.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('render with children snapshot', () => {
		const cmp = renderer.create(<TableWithHeader columns={columns}>
			{
				dataList.map(item => {
					return <TableTr key={item.id} data={item} columns={columns}/>
				})
			}
			</TableWithHeader>);
		const tree = cmp.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
