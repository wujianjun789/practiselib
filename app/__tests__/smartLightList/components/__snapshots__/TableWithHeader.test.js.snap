exports[`<TableWithHeader /> default render snapshot 1`] = `
<div
  className="table-responsive undefined">
  <table
    className="table table-hover">
    <thead>
      <tr>
        <th>
          设备名称
        </th>
        <th>
          value*10
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;

exports[`<TableWithHeader /> render with children snapshot 1`] = `
<div
  className="table-responsive undefined">
  <table
    className="table table-hover">
    <thead>
      <tr>
        <th>
          设备名称
        </th>
        <th>
          value*10
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className=""
        onClick={[Function]}>
        <td>
          test
        </td>
        <td>
          100
        </td>
      </tr>
      <tr
        className=""
        onClick={[Function]}>
        <td>
          test
        </td>
        <td>
          100
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`<TableWithHeader /> render with className="test" snapshot 1`] = `
<div
  className="table-responsive test">
  <table
    className="table table-hover">
    <thead>
      <tr>
        <th>
          设备名称
        </th>
        <th>
          value*10
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
